pkg/cmd/roachtest/roachtestutil/task/manager.go

```diff
@@ -73,9 +87,62 @@ func (m *manager) defaultOptions() []Option {
 	}
 }
 
+// Terminate will call the stop functions for every task started during the
+// test. Returns when all task functions have returned, or after a 5-minute
+// timeout, whichever comes first. If the timeout is reached, the function logs
+// a warning message and returns.
+func (m *manager) Terminate(l *logger.Logger) {
+	m.group.cancelAll()
+
+	doneCh := make(chan error)
+	go func() {
+		defer close(doneCh)
+		m.group.Wait()
+	}()
+
+	WaitForChannel(doneCh, "tasks", l)
+}
+
+// CompletedEvents returns a channel that will receive events for all tasks
+// started by the manager.
+func (m *manager) CompletedEvents() <-chan Event {
+	return m.events
+}
+
+// NewGroup creates a new group of tasks as a subgroup under the manager's
+// default group.
+func (m *manager) NewGroup(opts ...Option) Group {
+	return m.group.NewGroup(opts...)
+}
+
+// GoWithCancel runs GoWithCancel on the manager's default group.
 func (m *manager) GoWithCancel(fn Func, opts ...Option) context.CancelFunc {
-	opt := CombineOptions(OptionList(m.defaultOptions()...), OptionList(opts...))
-	groupCtx, cancel := context.WithCancel(m.ctx)
+	return m.group.GoWithCancel(fn, opts...)
+}
+
+// Go runs Go on the manager's default group.
+func (m *manager) Go(fn Func, opts ...Option) {
+	_ = m.group.GoWithCancel(fn, opts...)
+}
+
+func (t *group) NewGroup(opts ...Option) Group {
```

Comment:
For example, in the new unit test you added, if instead of closing the channels for each task you called `manger.Terminate()`, I would expect all the group tasks to get cancelled and unblock.

```
diff --git a/pkg/cmd/roachtest/roachtestutil/task/manager_test.go b/pkg/cmd/roachtest/roachtestutil/task/manager_test.go
index 54fe3838313..c7baf8a71de 100644
--- a/pkg/cmd/roachtest/roachtestutil/task/manager_test.go
+++ b/pkg/cmd/roachtest/roachtestutil/task/manager_test.go
@@ -148,7 +148,10 @@ func TestGroups(t *testing.T) {
        for i := 0; i < numTasks; i++ {
                channels[i] = make(chan struct{})
                g.Go(func(ctx context.Context, l *logger.Logger) error {
-                       <-channels[i]
+                       select {
+                       case <-ctx.Done():
+                       case <-channels[i]:
+                       }
                        return nil
                })
        }
@@ -162,16 +165,19 @@ func TestGroups(t *testing.T) {
 
        // Close channels one by one to complete all tasks, and ensure the group is
        // not done yet.
-       for i := 0; i < numTasks; i++ {
-               select {
-               case <-done:
-                       t.Fatal("group should not be done yet")
-               default:
-               }
-               // Close the channel and wait for the completed event.
-               close(channels[i])
-               <-m.CompletedEvents()
-       }
+       m.Terminate(nilLogger())

```

Reply:
Nice catch!
😱 Massive oversight while refactoring and no unit test to catch it!
I added a unit test now, that will test exactly this case.